google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/AddPerformanceMaxProductListingGroupTree.java

```diff
@@ -0,0 +1,556 @@
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param assetGroupId the asset group id for the Performance Max campaign.
+   * @param replaceExistingTree option to remove existing product tree from the passed in asset
+   *     group.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   */
+  private void runExample(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      long assetGroupId,
+      boolean replaceExistingTree)
+      throws Exception {
+    String assetGroupResourceName = ResourceNames.assetGroup(customerId, assetGroupId);
+
+    List<MutateOperation> operations = new ArrayList<>();
+
+    if (replaceExistingTree) {
+      List<AssetGroupListingGroupFilter> existingListingGroupFilters =
+          getAllExistingListingGroupFilterAssetsInAssetGroup(
+              googleAdsClient, customerId, assetGroupResourceName);
+
+      if (!existingListingGroupFilters.isEmpty()) {
+        // A special factory object that ensures the creation of remove operations in the
+        // correct order (child listing group filters must be removed before their parents).
+        AssetGroupListingGroupFilterRemoveOperationFactory removeOperationFactory =
+            new AssetGroupListingGroupFilterRemoveOperationFactory(existingListingGroupFilters);
+
+        operations.addAll(removeOperationFactory.removeAll());
+      }
+    }
+
+    // Uses a factory to create all the MutateOperations that manipulate a specific
+    // AssetGroup for a specific customer. The operations returned by the factory's methods
+    // are used to construct a new tree of filters. These filters can have parent-child
+    // relationships, and also include a special root that includes all children.
+    //
+    // When creating these filters, temporary IDs are used to create the hierarchy between
+    // each of the nodes in the tree, beginning with the root listing group filter.
+    //
+    // The factory created below is specific to a customerId and assetGroupId.
+    AssetGroupListingGroupFilterCreateOperationFactory createOperationFactory =
+        new AssetGroupListingGroupFilterCreateOperationFactory(
+            customerId, assetGroupId, TEMPORARY_ID_LISTING_GROUP_ROOT);
+
+    // Creates the operation to add the root node of the tree.
+    operations.add(createOperationFactory.createRoot());
+
+    // Creates an operation to add a leaf node for new products.
+    ListingGroupFilterDimension newProductDimension =
+        ListingGroupFilterDimension.newBuilder()
+            .setProductCondition(
+                ProductCondition.newBuilder()
+                    .setCondition(ListingGroupFilterProductCondition.NEW)
+                    .build())
+            .build();
+    operations.add(
+        createOperationFactory.createUnit(
+            TEMPORARY_ID_LISTING_GROUP_ROOT,
```

Comment:
realizing now that it might have been more elegant to save the listing group root in the factory.